### ISSUE TITLE
Hang possibly related to GC in preview7

### ISSUE THREAD

**Issue by GSPP** [dotnet/coreclr#25917]
The following program is expected to consume 1 full CPU core forever:

            byte[] encryptionData = new byte[16];

            using (var algorithm = new AesManaged())
            {
                algorithm.GenerateKey();
                algorithm.GenerateIV();
                algorithm.Mode = CipherMode.ECB;
                algorithm.Padding = PaddingMode.None;

                var cryptoTransform = algorithm.CreateEncryptor();

                while (true)
                {
                    var result = cryptoTransform.TransformFinalBlock(encryptionData, 0, encryptionData.Length);
                }
            }

But sometimes CPU usage stops. The debugger cannot stop the program at that point. Using Process Explorer I captured this stack for the main thread:

![image](https://user-images.githubusercontent.com/12032350/62006383-5692df00-b140-11e9-821c-8622a3453514.png)

![image](https://user-images.githubusercontent.com/12032350/62006397-75917100-b140-11e9-8290-b40f2390d9c3.png)

This is Windows 10 and a very recent `dbghelp.dll` so I don't understand why the stack is broken.

Reproducing this is a bit brittle. Sometimes, it is enough to start the program and the hang occurs after at most a minute or so. Sometimes, this does not seem to work. But starting PerfView and starting to collect data always leads to a hang within seconds.

**Comment by Maoni0** [comment]
@sywhang could you please take a look?

**Comment by sywhang** [comment]
@GSPP What version of CoreCLR are you using?

**Comment by GSPP** [comment]
Oh, I forgot to post the version.  I'm using the official preview7 (3.0.100-preview7-012821).

**Comment by GSPP** [comment]
I notice that whenever `IL_Throw` is on the stack the next stack frame is this `0x0000000` number in Process Explorer:

![image](https://user-images.githubusercontent.com/12032350/62141026-36078800-b2ec-11e9-93ee-cf851adc9ad3.png)

This is an exception heavy program. PerfView is blind to the time spent with exception processing:

![image](https://user-images.githubusercontent.com/12032350/62141302-a0b8c380-b2ec-11e9-943c-d1c23ccb28b7.png)

The difference between the numbers in red circles is exception processing.

Is something wrong with stack walking?

**Comment by GSPP** [comment]
I'm seeing the same stack walking issue with `JIT_NewArr1`. Also, I can reproduce the hang with an entirely different program.

![image](https://user-images.githubusercontent.com/12032350/62271674-7e808c00-b439-11e9-9542-8683a86342db.png)

Here are all other stacks in the currently hung process:

![image](https://user-images.githubusercontent.com/12032350/62271769-b982bf80-b439-11e9-91f0-6c256d516c70.png)

![image](https://user-images.githubusercontent.com/12032350/62271779-c30c2780-b439-11e9-869c-2390c70ac1ac.png)

![image](https://user-images.githubusercontent.com/12032350/62271798-c9020880-b439-11e9-9537-e49ca07b4cd3.png)

![image](https://user-images.githubusercontent.com/12032350/62271808-cf908000-b439-11e9-93fc-1b0735c1f46a.png)

![image](https://user-images.githubusercontent.com/12032350/62271816-d5866100-b439-11e9-9c24-ed42cfd7f257.png)

![image](https://user-images.githubusercontent.com/12032350/62271825-dd460580-b439-11e9-8a83-0422b083ab0c.png)

**Comment by GSPP** [comment]
I reported the strange stacks here: https://github.com/dotnet/coreclr/issues/26034

**Comment by sywhang** [comment]
Hi @GSPP,

Sorry it took a while to get back. Unfortunately I can't repro the hang on any of the platforms I've tried so far (Windows 10 - Core 3.0 Preview 5, 6, and 7). Initially I thought the issue might be related to some code path I changed in the `UpdatePostGCCounters` since that's where the exception is thrown but it seems like the stack you have in the original description of the issue is very suspicious looking because the offsets are all off. It looks like a corrupted stack..

Would it be possible for you to capture a dump using a native debugger like WinDBG and share the dump with me? If you'd like to provide it privately instead of posting on GitHub you can reach out to me at suwhang AT microsoft.com. 

Thanks!

**Comment by GSPP** [comment]
I have created a small repro solution which I am attaching. I also captured a dump that I sent to you by email.

I found that server GC makes the repro go away.

Reproducing it is brittle. It happens more often in Debug mode it seems. The act of starting PerfView (not even starting to collect) triggers it quite often but I also can reproduce without PerfView. I captured a dump in Release mode without PerfView called "ExceptionStackRepro GC Release NoPerfView.dmp". Likely, this is the dump to look at first.

The same solution contains a repro for the stack walking issue. There's a dump for that, too.

[ExceptionStackRepro.zip](https://github.com/dotnet/coreclr/files/3475789/ExceptionStackRepro.zip)

**Comment by sywhang** [comment]
Thanks for the dump @GSPP! I found the issue that causes the hang so I'll be sending out a fix for that tonight.

**Comment by sywhang** [comment]
Keeping it open for 3.0 port.

**Comment by GSPP** [comment]
Oh, it was a division by zero? I had programs crashing on me with that error in internal CLR code. I could no longer reproduce it.

I'd like to test the fix. I guess I wait for the port issue to close and the next nightly should have the fix, right?

**Comment by sywhang** [comment]
The fix already went into the master branch. I believe the nightly builds will take a few days to pick it up. You can try the fix right now though if you want - You can build CoreCLR repo from the master branch, and then copy over the binaries built over. You can publish your repro app to a self-contained app by using the command `dotnet publish -r win-x64` and then copy over all the dll files from the output directory of CoreCLR (bin\Product\Windows.x64.Debug\*.dll)

The port PR is for 3.0 release specifically. This is not the nightly build and it's for the preview 9 release of .NET Core 3.0.

**Comment by jeffschwMSFT** [comment]
Closing with #26085 and #26089

**Comment by GSPP** [comment]
Still reproducible with 3.0.0-preview9-19419-03.

**Comment by GSPP** [comment]
I'm unable to get the repro project to use the master version that I installed from https://github.com/dotnet/core-sdk#installers-and-binaries. I published like you advised. In Process Explorer I can see that `coreclr.dll` is still loaded from the shared `preview9` folder. I added `--self-contained` and set `DOTNET_MULTILEVEL_LOOKUP=0` which did not help.

I don't think I have any machine set up to build CoreCLR. Can I just use the master daily builds?

Any advice on how to make it work? I'd like to test the fix.
